include/silo/common/panic.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <fmt/format.h>
+
+namespace silo::common {
+
+void panic(const std::string& msg);
+
+/// Passes arguments to `fmt::format` (at least a format string
+/// argument is required), then if the `DEBUG` environment variable is
+/// set to anything else than the string "0", the resulting string is
+/// printed to stderr and `abort` is called, otherwise a
+/// `std::runtime_error` is thrown (details subject to
+/// change). `PANIC` should only be used in situations that can never
+/// occur unless there is a bug; the aim, besides allowing the
+/// programmer to declare clearly what constitutes buggy behaviour, is
+/// to allow these situations to be debugged via a debugger like gdb
+/// without needing contortions like setting breakpoints, or also to
+/// allow the collection of core dumps of these situations.
+#define PANIC(...)                                                                            \
+   do {                                                                                       \
+      auto __panic_msg = fmt::format(__VA_ARGS__);                                            \
+      auto __panic_msg1 = fmt::format("PANIC: {} at {}:{}", __panic_msg, __FILE__, __LINE__); \
+      silo::common::panic(__panic_msg1);                                                      \
+   } while (0)
```

Comment:
Out of curiosity: Why is there a `do while` loop?

Reply:
To make the macro usable correctly in all contexts. It's a workaround for one of the CPP macro dangers. (Rust macros don't have this problem.)

https://stackoverflow.com/questions/1067226/c-multi-line-macro-do-while0-vs-scope-block

Reply:
Interesting. Thanks :)